### PR TITLE
LOG-2204: Modify routemaster-blackhole to push all events received to data-sink

### PR DESCRIPTION
**Pull request by pedrocunha**
===

Jira story [#LOG-2204](https://deliveroo.atlassian.net/browse/LOG-2204) in project *Logistics*:

> Investigate if we can modify routemaster-blackhole to send all events received to data-sink. Then create a PLAYBOOK that explains how to import data to snowflake, and query by window/topic and select the urls/events/ids published
> 
> Also consider this app to subscribe by default to all topics currently in place (today it only listens to topics set on an ENV variable)

### CODE REVIEW

File: config.ru [hunk 1]
```diff
@@ -4,16 +4,39 @@ require 'routemaster/receiver'
...
     end
+    send_to_data_sink(events)
+  end
+
+  private
+
+  def send_to_data_sink(events)
+    return unless data_sink_enabled?
+    events.map do |event|
+      Thread.new do
```
Comment (mjkillough) [inline config.ru]: We should use a thread-pool to ensure we never try to create a thread per event for large batches.

I know the `parallel` gem would make this really easy, but maybe we already have a favourite thread-pool gem.

**Review by mjkillough — APPROVED** [review]

File: config.ru (hunk 1 above)
Comment (pedrocunha) [reply config.ru]: @mjkillough but here we are creating a thread per event. Are you worried about spawning too much threads per process?

Comment (mjkillough) [reply config.ru]: Yeah.

The default batch size for Routemaster is 100, which is more threads than I'd normally start in one go. I would prefer to limit the number of threads we create, but I don't think it'd really be disastrous to leave it as-is. This service doesn't do anything business critical.

Comment (me) [reply config.ru]: https://github.com/ruby-concurrency/concurrent-ruby/blob/master/doc/thread_pools.md is also nice (we're using it in data-sink)

File: data_sink_client.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+require 'oj'
+
+class DataSinkClient
+  include Singleton
```
Comment (me) [inline data_sink_client.rb]: Why not class methods?

**Review by me — APPROVED** [review]
👍

File: data_sink_client.rb (hunk 2 above)
Comment (pedrocunha) [reply data_sink_client.rb]: I think this way is more concise for this particular usecase where I do not want to instance a client multiple times

File: config.ru [hunk 3]
```diff
@@ -4,16 +4,41 @@ require 'routemaster/receiver'
...
+    events.map do |event|
+      pool = Concurrent::FixedThreadPool.new(Concurrent.processor_count)
+      pool.post do
+        event = event.merge(
+          subscriber: SUBSCRIBER_NAME,
+          received_at: (Time.now.utc.to_f * 1e3).to_i
+        )
+        DataSinkClient.instance.post(event)
+      end
+    end.map(&:join)
```
Comment (mjkillough) [inline config.ru]: I think this makes `Concurrent.processor_count` threads per event. 😃 

I think you'll also need to change the `.map(&:join)`. It seems `FixedThreadPool` has a `#wait_for_termination`.

Comment (pedrocunha) [reply config.ru]: I wanted to try this locally first and pushed to remote by mistake. Forget everyone gets an email 😅

Comment (mjkillough) [reply config.ru]: Aha, I thought that might be the case. :) I wouldn't have mentioned anything had the PR not still been 'approved'.

**Comment by mjkillough** [comment]
Awesome, LGTM! 👍 Thanks a lot.

**Comment by Antti** [comment]
@pedrocunha Is it not a bit too expensive to start a new thread pool each time we get a request? Would it make more sense to keep the thread pool running?

**Comment by pedrocunha** [comment]
Do you think that could have caused throughput problems? I guess with Unicorn is probably safe to keep a thread pool running. Another issue on the current architecture is that if DataSink slows down it can cause every web request to slow down. So we could potentially introduce Sidekiq

**Pull request merged**